src/community/wmts-multi-dimensional/src/main/java/org/geoserver/gwc/wmts/dimensions/DimensionsUtils.java

```diff
@@ -0,0 +1,159 @@
+/* (c) 2016 Open Source Geospatial Foundation - all rights reserved
+ * This code is licensed under the GPL 2.0 license, available at the root
+ * application directory.
+ */
+package org.geoserver.gwc.wmts.dimensions;
+
+import org.geoserver.catalog.*;
+import org.geoserver.gwc.wmts.Tuple;
+import org.geoserver.util.ISO8601Formatter;
+import org.geoserver.wms.WMS;
+import org.geotools.util.Converters;
+import org.geotools.util.Range;
+
+import java.io.Serializable;
+import java.util.*;
+
+/**
+ * Some utils methods useful to interact with dimensions.
+ */
+public final class DimensionsUtils {
+
+    /**
+     * Helper method that will extract a layer dimensions.
+     */
+    public static List<Dimension> extractDimensions(WMS wms, LayerInfo layerInfo) {
+        ResourceInfo resourceInfo = layerInfo.getResource();
+        if (resourceInfo instanceof FeatureTypeInfo) {
+            return extractDimensions(wms, layerInfo, (FeatureTypeInfo) resourceInfo);
+        }
+        if (resourceInfo instanceof CoverageInfo) {
+            return extractDimensions(wms, layerInfo, (CoverageInfo) resourceInfo);
+        }
+        return Collections.emptyList();
+    }
+
+    /**
+     * Helper method that will extract the dimensions from a feature type info.
+     */
+    private static List<Dimension> extractDimensions(WMS wms, LayerInfo layerInfo, FeatureTypeInfo typeInfo) {
+        List<Dimension> dimensions = new ArrayList<>();
+        DimensionInfo timeDimension = typeInfo.getMetadata().get(ResourceInfo.TIME, DimensionInfo.class);
+        if (timeDimension != null) {
+            dimensions.add(new VectorTimeDimension(wms, layerInfo, timeDimension));
+        }
+        DimensionInfo elevationDimension = typeInfo.getMetadata().get(ResourceInfo.ELEVATION, DimensionInfo.class);
+        if (elevationDimension != null) {
+            dimensions.add(new VectorElevationDimension(wms, layerInfo, elevationDimension));
+        }
+        return dimensions;
+    }
+
+    /**
+     * Helper method that will extract the dimensions from a coverage type info.
+     */
+    private static List<Dimension> extractDimensions(WMS wms, LayerInfo layerInfo, CoverageInfo typeInfo) {
+        List<Dimension> dimensions = new ArrayList<>();
+        for (Map.Entry<String, Serializable> entry : typeInfo.getMetadata().entrySet()) {
+            String key = entry.getKey();
+            Serializable value = entry.getValue();
+            if (key.equals(ResourceInfo.TIME)) {
+                dimensions.add(new RasterTimeDimension(wms, layerInfo, Converters.convert(value, DimensionInfo.class)));
```

Comment:
The code should check if the DimensionInfo is enabled. We can have layers that have had a dimension configured and no more enabled, but keep the settings to make sure we don't lose the edits.
In particular I have a layer that used to have a time dimension, that then got removed at the source level, and disabled in the UI, the current code tries to extract a domain that's not there causing a NPE:

````
05 set 12:36:34 ERROR [geowebcache.GeoWebCacheDispatcher] - null http://localhost:8080/geoserver/gwc/service/wmts
java.lang.NullPointerException
	at org.geoserver.catalog.util.ReaderDimensionsAccessor.getTimeDomain(ReaderDimensionsAccessor.java:152)
	at org.geoserver.gwc.wmts.dimensions.RasterTimeDimension.getDomainValues(RasterTimeDimension.java:37)
	at org.geoserver.gwc.wmts.dimensions.Dimension.getDomainValuesAsStrings(Dimension.java:180)
	at org.geoserver.gwc.wmts.MultiDimensionalExtension.encodeLayerDimension(MultiDimensionalExtension.java:226)
	at org.geoserver.gwc.wmts.MultiDimensionalExtension.encodeLayerDimensions(MultiDimensionalExtension.java:212)
	at org.geoserver.gwc.wmts.MultiDimensionalExtension.encodeLayer(MultiDimensionalExtension.java:137)
	at org.geowebcache.service.wmts.WMTSGetCapabilities.layer(WMTSGetCapabilities.java:453)
	at org.geowebcache.service.wmts.WMTSGetCapabilities.contents(WMTSGetCapabilities.java:395)
	at org.geowebcache.service.wmts.WMTSGetCapabilities.generateGetCapabilities(WMTSGetCapabilities.java:146)
	at org.geowebcache.service.wmts.WMTSGetCapabilities.writeResponse(WMTSGetCapabilities.java:90)
	at org.geowebcache.service.wmts.WMTSService.handleRequest(WMTSService.java:315)
	at org.geowebcache.service.wmts.WMTSService$$FastClassBySpringCGLIB$$e221a779.invoke(<generated>)
````